saq/worker.py

```diff
@@ -307,22 +327,38 @@ async def process(self) -> bool:
             job.started = now()
             job.attempts += 1
             await job.update(status=Status.ACTIVE)
-            context = t.cast(CtxType, {**self.context, "job": job})
+            context = {**self.context, "job": job}
             await self._before_process(context)
             logger.info("Processing %s", job.info(logger.isEnabledFor(logging.DEBUG)))
 
-            function = ensure_coroutine_function(self.functions[job.function], self.pool)
+            function = self.functions.get(job.function)
+            assert function is not None
+            function = ensure_coroutine_function(function, self.pool)
             task = asyncio.create_task(function(context, **(job.kwargs or {})))
-            self.job_task_contexts[job] = {"task": task, "aborted": None}
-            result = await asyncio.wait_for(task, job.timeout if job.timeout else None)
+            self.job_task_contexts[job] = JobTaskContext(task=task, aborted=None)
+            try:
+                result = await asyncio.wait_for(
```

Comment:
why did this flow have to change?

Reply:
We need an asyncio shield over the future we pass to `wait_for` to stop the cancellation signal from reaching the task. But it has another consequence in that the `wait_for` timeout also doesn't send the cancel signal that it previously did, so we need both the shield as well as explicit cancellation on timeout.